# Trabalho Prático 2 - Temas de Álgebra

### Mestrado em Matemática e Computação, Ano Letivo 2023-24

-------------------------------------

#### Enunciado do trabalho

Considere uma matriz A de característica máxima, 4x3, e b um vector 4x1, aleatórios. Assuma que car(A)<car[A | b], ou de forma equivalente, que det([A | b]) != 0.

Pode, por exemplo, usar
https://sagecell.sagemath.org/?z=eJxzVLBVKErMS8nPjc9NLCnKrNCIitIx0THW5OVKwi5lCJQqKMrMK9FISS3RSMrJT86GyRvqGOlEO-ooJMVqairaGoAVppaUVMZD1INkgGI5iSWpFRqOmjoKEFaSJgDfnSj3&lang=sage&interacts=eJyLjgUAARUAuQ==

(várias vezes, se não lhe agradar o resultado obtido)

Use reflexões elementares ou rotações de Givens para encontrar uma base ortonormada do espaço das colunas de A.
Usando a base ortonormada obtida, encontre a solução no sentido dos mínimos quadrados de Ax=b.

------------------------------------------------

### Resolução do trabalho

Referências webgráficas:

- https://www.ime.unicamp.br/~marcia/AlgebraLinear/fat_QR%28householder%29.html
- https://www.ime.unicamp.br/~marcia/AlgebraLinear/quadrados_minimos.html
- https://pages.stat.wisc.edu/~bwu62/771/businger1965.pdf

Geram-se duas matriz aleatória A e b:

In [1]:
A = random_matrix(ZZ,4,3)
b = random_matrix(ZZ,4,1)
print(det(block_matrix(1,2,[A, b]))!=0)
pretty_print(A, b)
latex(A), latex(b)

True


[-1 -4  2]
[ 1  0  0]
[-2  3 -1]
[ 1  0  1] [-1]
[ 0]
[ 1]
[24]

(\left(\begin{array}{rrr}
 -1 & -4 & 2 \\
 1 & 0 & 0 \\
 -2 & 3 & -1 \\
 1 & 0 & 1
 \end{array}\right),
 \left(\begin{array}{r}
 -1 \\
 0 \\
 1 \\
 24
 \end{array}\right))

In [2]:
# Característica da matriz A (máximo 4x3)
rank(A)

3

A matriz $A$ tem característica igual a 3 (equivalente ao seu valor máximo), com 3 vetores/colunas no espaço de A.

In [3]:
Ab = block_matrix(1,2,[A, b])
pretty_print(Ab)

[-1 -4  2|-1]
[ 1  0  0| 0]
[-2  3 -1| 1]
[ 1  0  1|24]

In [4]:
rank(Ab)

4

In [5]:
det(Ab)

47

Verifica-se que $car(A)<car[A | b]$ e que $det([A | b]) != 0$.

Agora, vamos aplicar __reflexões elementares__ (transformações de Householder) com o objetivo de obter uma base ortonormada do espaço das colunas de A (originando uma fatorização $A = QR$), para que no fim se possa resolver a equação $Ax=b$ através da técnica dos mínimos quadrados. Como a matriz A possui dimensão 4x3, ir-se-ão aplicar 3 vezes as rotações elementares.

In [6]:
mu = 1

# matriz identidade de tamanho igual ao compriemnto de cada coluna/vetor do espaço de A
I4 = identity_matrix(4)

- Cálculo da refelxão elementar para a 1ª coluna da matriz A

In [7]:
a1 = A[:,0]
a1

[-1]
[ 1]
[-2]
[ 1]

In [8]:
e1 = I4[:,0]
e1

[1]
[0]
[0]
[0]

In [9]:
norma_a1 = a1.norm()
norma_a1

2.6457513110645907

In [10]:
u1 = a1+norma_a1*e1

# Primeira matriz de reflexão
R1= I4-2*u1*transpose(u1)/(norm(u1)**2)
pretty_print(u1, R1)

[1.6457513110645907]
[               1.0]
[              -2.0]
[               1.0] [  0.3779644730092271 -0.37796447300922736   0.7559289460184547 -0.37796447300922736]
[-0.37796447300922736   0.7703392544984621  0.45932149100307584 -0.22966074550153792]
[  0.7559289460184547  0.45932149100307584  0.08135701799384831  0.45932149100307584]
[-0.37796447300922736 -0.22966074550153792  0.45932149100307584   0.7703392544984621]

In [11]:
A2=R1*A
pretty_print(A2)

[     -2.645751311064591      0.7559289460184558     -0.3779644730092279]
[-1.1102230246251565e-16       2.889822365046137     -1.4449111825230685]
[ 3.3306690738754696e-16      -2.779644730092274       1.889822365046137]
[-1.1102230246251565e-16       2.889822365046137     -0.4449111825230686]

- Cálculos das reflexões elementares para a primeira coluna da matriz $A2sub$

In [12]:
A2sub = A2[1:4, 1:4]
A2sub

[  2.889822365046137 -1.4449111825230685]
[ -2.779644730092274   1.889822365046137]
[  2.889822365046137 -0.4449111825230686]

In [13]:
a2 = A2sub[:,0]
a2

[ 2.889822365046137]
[-2.779644730092274]
[ 2.889822365046137]

In [14]:
norma_a2 = a2.norm()
norma_a2

4.9425268262874855

In [15]:
# matriz identidade
I3 = identity_matrix(3)

u2 = a2-norm(a2)*I3[:,0]
u2

[-2.0527044612413485]
[ -2.779644730092274]
[  2.889822365046137]

In [16]:
R2sub = I3-2*u2*transpose(u2)/(norm(u2)**2)
pretty_print(R2sub)

[ 0.5846852160065643 -0.5623934533462446  0.5846852160065642]
[-0.5623934533462446  0.2384417589822332  0.7917443597568246]
[ 0.5846852160065642  0.7917443597568246  0.1768730250112024]

In [17]:
R2sub*A2sub

[     4.942526826287486     -2.167774923810301]
[ 6.402121107990518e-16     0.9109652392479406]
[-7.194965608196698e-16      0.572745204889615]

In [18]:
# Segunda matriz de reflexão
R2 = block_diagonal_matrix(matrix([[1]]),R2sub)
pretty_print(R2)

[                1.0|                0.0                 0.0                 0.0]
[-------------------+-----------------------------------------------------------]
[                0.0| 0.5846852160065643 -0.5623934533462446  0.5846852160065642]
[                0.0|-0.5623934533462446  0.2384417589822332  0.7917443597568246]
[                0.0| 0.5846852160065642  0.7917443597568246  0.1768730250112024]

In [19]:
A3 = R2sub*A2sub
pretty_print(A3)

[     4.942526826287486     -2.167774923810301]
[ 6.402121107990518e-16     0.9109652392479406]
[-7.194965608196698e-16      0.572745204889615]

- Cálculos da reflexão elementar para a primeira coluna da matriz $A3sub$

In [20]:
A3sub = A3[1:3,1:3]
A3sub

[0.9109652392479406]
[ 0.572745204889615]

In [21]:
a3 = A3sub
a3

[0.9109652392479406]
[ 0.572745204889615]

In [22]:
# Matriz identidade
I2 = identity_matrix(2)

e3 = I2[:,0]
e3

[1]
[0]

In [23]:
norma_a3=a3.norm()
norma_a3

1.0760551736979405

In [24]:
u3=a3+norma_a3*e3
u3

[1.987020412945881]
[0.572745204889615]

In [25]:
R3sub = I2-2*u3*transpose(u3)/(u3.norm()**2)
R3sub

[-0.8465785598310387  -0.532263789895954]
[ -0.532263789895954  0.8465785598310387]

In [26]:
aux = matrix(QQ,[[1,0],[0,1]])

# Terceira matriz de reflexão
R3 = block_diagonal_matrix(aux,R3sub)
pretty_print(R3)

[                1.0                 0.0|                0.0                 0.0]
[                0.0                 1.0|                0.0                 0.0]
[---------------------------------------+---------------------------------------]
[                0.0                 0.0|-0.8465785598310387  -0.532263789895954]
[                0.0                 0.0| -0.532263789895954  0.8465785598310387]

- Obter matriz $Q$, que contém a base ortonormada da matriz A

In [27]:
Q_t = R3*(R2*R1)
Q_t

[  0.3779644730092271 -0.37796447300922736   0.7559289460184547 -0.37796447300922736]
[ -0.8671099695241202   0.0578073313016079   0.4913623160636682  0.05780733130160779]
[-0.24455799402225922 0.016303866268150134  -0.3260773253630124  -0.9130165110164344]
[ 0.21320071635561053   0.9238697708743119  0.28426762180748033  -0.1421338109037407]

In [28]:
Q = Q_t.transpose()
pretty_print(Q)

[  0.3779644730092271  -0.8671099695241202 -0.24455799402225922  0.21320071635561053]
[-0.37796447300922736   0.0578073313016079 0.016303866268150134   0.9238697708743119]
[  0.7559289460184547   0.4913623160636682  -0.3260773253630124  0.28426762180748033]
[-0.37796447300922736  0.05780733130160779  -0.9130165110164344  -0.1421338109037407]

- Obter matriz $R$, com a sua forma __triangular superior__ nas primeiras 3 linhas da matriz

In [29]:
R = R3*(R2*(R1*A))
pretty_print(R)

[     -2.645751311064591      0.7559289460184558     -0.3779644730092279]
[ -3.171408460347205e-16       4.942526826287486      -2.167774923810301]
[-1.4103342580355948e-16 -1.5902788046766843e-16     -1.0760551736979405]
[ 1.2295009014781147e-16   -9.49872086693293e-16  1.1613693984590276e-16]

Apesar de nos valores inferiores à zona triangular superior da matriz $R$ não serem exatamnete iguais a zero, estes são muito próximos de zero (ordem de grandeza extremamente baixa), podendo-se afirmar que a matriz $R$ obtida é válida.

Assim, obtemos uma fatorização de $A = QR$.

In [30]:
verify_A = round(Q*R) # igual à matriz original A
pretty_print(verify_A)

[-1.0 -4.0  2.0]
[ 1.0 -0.0  0.0]
[-2.0  3.0 -1.0]
[ 1.0 -0.0  1.0]

- Resolução da equação $Ax=b$ através dos mínimos quadrados

Para resolver a equação $Ax=b$ pelo método dos mínimos quadrados, precisamos encontrar a solução $x=(x_1,x_2,x_3)$ que minimiza a norma do resíduo $||b-Ax||$. Podemos usar os resultados obtidos com a fatorização QR de A, nomeadmente as matrizes $Q^T$ e $R$, para obter a solução.

Obtemos a matriz triangular superior de $R$, eliminando a última linha da matriz, de forma a garantir que a matriz seja quadrada. Sabendo que $c=Q^T b$, aplica-se algo semelhança ao passo anterior no vetor $c$ até os 3 primeiros valores do vetor ($c_3$). Por fim, a solução da equação $Ax=b$ pelo método dos mínimos quadrados é equivalente à solução $x = R^{-1}_{triangular} c_3$, já que minimizar o resíduo $||b-Ax||$ é é o mesmo que minimizar $||c-QAx||$.


In [31]:
c = Q_t*b
c

[ -8.69318287921223]
[2.7458482368263755]
[-21.99391559573518]
[-3.340144556237907]

In [32]:
# apenas com a sua forma triangular
R_triangular = R[:3]
R_triangular

[     -2.645751311064591      0.7559289460184558     -0.3779644730092279]
[ -3.171408460347205e-16       4.942526826287486      -2.167774923810301]
[-1.4103342580355948e-16 -1.5902788046766843e-16     -1.0760551736979405]

In [33]:
# 3 primeiros valores do vetor c
c_3 = c[:3]
c_3

[ -8.69318287921223]
[2.7458482368263755]
[-21.99391559573518]

In [34]:
x = (R_triangular.inverse())*c_3
x

[3.0858585858585856]
[ 9.520202020202019]
[ 20.43939393939394]

Então, pelo método dos mínimos quadrados, usando reflexões elementares (trasnformações de Householder), a solução da equação $Ax=b$ é igual a:

- $x_1=3.0858585858585856$
- $x_2=9.520202020202019$
- $x_3=20.43939393939394$